controller.go

```diff
@@ -266,7 +270,27 @@ func updateAlarmDefinition(id string, r Resource) error {
 	definitionRequest := convertToADRequest(r.Spec.AlarmDefinitionElement)
 	result, err := monascaclient.PatchAlarmDefinition(id, definitionRequest)
 	if err != nil {
-		return err
+		if !strings.HasPrefix(err.Error(), "Error: 422") {
+			return err
+		}
+		// Some updates are not allowed, so try deleting and recreating instead
+		// NOTE: this will remove the history from the alarm definition
+		log.Printf("Failed to update alarm %s, attempting delete and recreate", id)
+		deleteErr := removeAlarmDefinition(id, r.Spec.AlarmDefinitionElement)
+		if deleteErr != nil {
+			log.Printf("Error deleting definition: %s", deleteErr.Error())
+			// return the original error
+			return err
+		}
+		// Remove the ID from the k8s resource to reduce
+		// confusion if the create fails
+		emptyID := make(map[string]map[string]string)
+		emptyID["alarmDefinitionSpec"] = make(map[string]string)
+		emptyID["alarmDefinitionSpec"]["id"] = ""
+		patchResource(r, emptyID)
```

Comment:
This is the only place you don't check the error code and log it if it fails

Reply:
It actually doesn't change any behavior if this errors, it's done more to avoid confusion on the user side.